packages/flutter/lib/src/material/app_bar.dart

```diff
@@ -24,6 +24,65 @@ abstract class AppBarBottomWidget extends Widget {
   double get bottomHeight;
 }
 
+enum _ToolBarSlot {
+  leading,
+  title,
+  actions,
+}
+
+class _ToolBarLayout extends MultiChildLayoutDelegate {
+  _ToolBarLayout({ this.centerTitle });
+
+  // If false the title should be left or right justified within the space bewteen
+  // the leading and actions widgets, depending on the locale's writing direction.
+  // If true the title is centered within the toolbar (not within the horizontal
+  // space bewteen the leading and actions widgets).
+  final bool centerTitle;
+
+  static const double kLeadingWidth = 24.0;
+  static const double kTitleLeft = 64.0; // The AppBar pads left and right an additional 8.0.
+
+  @override
+  void performLayout(Size size) {
+    double actionsWidth = 0.0;
+
+    if (hasChild(_ToolBarSlot.leading)) {
+      final BoxConstraints constraints = new BoxConstraints.tight(new Size(kLeadingWidth, size.height));
+      layoutChild(_ToolBarSlot.leading, constraints);
+      positionChild(_ToolBarSlot.leading, Offset.zero);
+    }
+
+    if (hasChild(_ToolBarSlot.actions)) {
+      final BoxConstraints constraints = new BoxConstraints.loose(size);
+      actionsWidth = layoutChild(_ToolBarSlot.actions, constraints).width;
+      positionChild(_ToolBarSlot.actions, new Offset(size.width - actionsWidth, 0.0));
+    }
+
+    if (hasChild(_ToolBarSlot.title)) {
+      final double maxWidth = size.width - kTitleLeft - actionsWidth;
+      final BoxConstraints constraints = new BoxConstraints.loose(size).copyWith(maxWidth: maxWidth);
+      final Size titleSize = layoutChild(_ToolBarSlot.title, constraints);
+      final double titleY = (size.height - titleSize.height) / 2.0;
+      double titleX = kTitleLeft;
+
+      // If the centered title will not fit between the leading and actions
+      // widgets, then justify it with the wider one.
```

Comment:
The title is constrained to be no wider than the available space between the leading and actions widgets. If it doesn't fit when centered then it will overlap the wider of the leading and actions widgets. It can't overlap both because [see the first sentence].

So when the centered title doesn't fit, either its left edge is set to the leading widget's right edge, or its right edge is set to the actions widgets' left edge.  That's what I mean by "justified". In this case the code justifies the overlapping title edge.